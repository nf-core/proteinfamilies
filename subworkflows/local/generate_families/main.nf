/*
    FAMILY MODEL GENERATION
*/

include { ALIGN_SEQUENCES  } from '../../../subworkflows/local/align_sequences'
include { CLIPKIT          } from '../../../modules/nf-core/clipkit/main'
include { HMMER_HMMBUILD   } from '../../../modules/nf-core/hmmer/hmmbuild/main'
include { HMMER_HMMSEARCH  } from '../../../modules/nf-core/hmmer/hmmsearch/main'
include { FILTER_RECRUITED } from '../../../modules/local/filter_recruited/main'
include { HMMER_HMMALIGN   } from '../../../modules/nf-core/hmmer/hmmalign/main'

workflow GENERATE_FAMILIES {
    take:
    sequences // tuple val(meta), path(fasta)
    fasta_chunks

    main:
    ch_versions = Channel.empty()
    ch_msa      = Channel.empty()
    ch_fasta    = Channel.empty()
    ch_hmm      = Channel.empty()

    ch_fasta = fasta_chunks
        .transpose()
        .map { meta, file_path ->
            [ [id: meta.id, chunk: file(file_path, checkIfExists: true).baseName], file_path ]
        }

    ALIGN_SEQUENCES( ch_fasta, params.alignment_tool )
    ch_versions = ch_versions.mix( ALIGN_SEQUENCES.out.versions )
    ch_msa = ALIGN_SEQUENCES.out.alignments

    if (params.trim_msa) {
        CLIPKIT( ch_msa, params.clipkit_out_format )
        ch_versions = ch_versions.mix( CLIPKIT.out.versions )
        ch_msa = CLIPKIT.out.clipkit
    }

    HMMER_HMMBUILD( ch_msa, [] )
    ch_versions = ch_versions.mix( HMMER_HMMBUILD.out.versions )
    ch_hmm = HMMER_HMMBUILD.out.hmm

    // Combine with same id to ensure in sync
    ch_input_for_hmmsearch = ch_hmm
        .map { meta, hmm -> [ [id: meta.id], meta, hmm ] }
        .combine(sequences, by: 0)
        .map { id, meta, hmm, seqs -> [ meta, hmm, seqs, false, params.hmmsearch_write_target, params.hmmsearch_write_domain ] }

    if (params.recruit_sequences_with_models) {
        HMMER_HMMSEARCH( ch_input_for_hmmsearch )
        ch_versions = ch_versions.mix( HMMER_HMMSEARCH.out.versions )

        // Combine with same id to ensure in sync
        ch_input_for_filter_recruited = HMMER_HMMSEARCH.out.domain_summary
            .map { meta, domtbl -> [ [id: meta.id], meta, domtbl ] }
            .combine(sequences, by: 0)
            .map { id, meta, domtbl, seqs -> [ meta, domtbl, seqs ] }

        FILTER_RECRUITED( ch_input_for_filter_recruited, params.hmmsearch_query_length_threshold )
        ch_versions = ch_versions.mix( FILTER_RECRUITED.out.versions )
        ch_fasta = FILTER_RECRUITED.out.fasta

        // Join to ensure in sync
        ch_input_for_hmmalign = ch_fasta
            .join(ch_hmm)
            .multiMap { meta, seqs, hmms ->
                seq: [ meta, seqs ]
                hmm: [ hmms ]
            }

        HMMER_HMMALIGN( ch_input_for_hmmalign.seq, ch_input_for_hmmalign.hmm )
        ch_versions = ch_versions.mix( HMMER_HMMALIGN.out.versions )
        ch_msa = HMMER_HMMALIGN.out.sto
    }

    emit:
    versions = ch_versions
    msa      = ch_msa
    fasta    = ch_fasta
    hmm      = ch_hmm
}
