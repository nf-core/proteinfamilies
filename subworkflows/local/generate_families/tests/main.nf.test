nextflow_workflow {

    name "Test Subworkflow GENERATE_FAMILIES"
    script "../main.nf"
    workflow "GENERATE_FAMILIES"

    test("fasta") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test_minimal' ],
                    file(params.pipelines_testdata_base_path + 'test_data/mgnifams_input_small.fa', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id: 'test_chunks' ],
                    [
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_1.fasta', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_2.fasta', checkIfExists: true)
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    file(workflow.out.hmm[0][1]).name,
                    workflow.out.seed_msa,
                    workflow.out.full_msa,
                    workflow.out.fasta,
                    workflow.out.versions).match()}
            )
        }
    }

}
