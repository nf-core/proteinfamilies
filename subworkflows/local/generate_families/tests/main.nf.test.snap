{
    "fasta": {
        "content": [
            "test_chunk_1.hmm.gz",
            [
                [
                    {
                        "id": "test",
                        "chunk": "chunk_1"
                    },
                    "test_chunk_1.clipkit:md5,dae9d5f940e2f43687e068155ebfe1ae"
                ],
                [
                    {
                        "id": "test",
                        "chunk": "chunk_2"
                    },
                    "test_chunk_2.clipkit:md5,adcd7fc302b4cd66c219d2df51a7e7e8"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "chunk": "chunk_1"
                    },
                    "test_chunk_1.sto.gz:md5,93d4f5b7b89ead091ef68dedd70ca2f5"
                ],
                [
                    {
                        "id": "test",
                        "chunk": "chunk_2"
                    },
                    "test_chunk_2.sto.gz:md5,e962e2d7e32fba8685d9670709a6835c"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "chunk": "chunk_1"
                    },
                    "test_chunk_1.fasta.gz:md5,5d14fec7a4e06606eda48f8c912323ab"
                ],
                [
                    {
                        "id": "test",
                        "chunk": "chunk_2"
                    },
                    "test_chunk_2.fasta.gz:md5,7ae62909eb81f92c331b2048a8ac34d3"
                ]
            ],
            [
                "versions.yml:md5,247ef08f0b089979a50b118c7fcdc2b0",
                "versions.yml:md5,247ef08f0b089979a50b118c7fcdc2b0",
                "versions.yml:md5,45eba8d30565dc253428be013fbd48cb",
                "versions.yml:md5,45eba8d30565dc253428be013fbd48cb",
                "versions.yml:md5,cf594456f2dd939c2364378f09c5d88a",
                "versions.yml:md5,cf594456f2dd939c2364378f09c5d88a",
                "versions.yml:md5,e57cf4916752bb72d16bd2e7072b9c26",
                "versions.yml:md5,e57cf4916752bb72d16bd2e7072b9c26",
                "versions.yml:md5,ef2d0b048087d63b5fcac1c32e704791",
                "versions.yml:md5,ef2d0b048087d63b5fcac1c32e704791",
                "versions.yml:md5,fc501370ac8ce03b2fcc4377984702f1",
                "versions.yml:md5,fc501370ac8ce03b2fcc4377984702f1"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-06-13T08:10:49.910402535"
    }
}